### PR TITLE
Psr 29 polysync echo command line tool

### PR DESCRIPTION
**Pull request by abstractmachines**
Classes PolySyncGetOpt and PolySyncEchoHelp added for modularity.

### CODE REVIEW

File: Echo/CMakeLists.txt [hunk 1]
```diff
@@ -15,6 +15,7 @@ set( PSYNC_HOME
     )
```
Comment (shnewto) [inline Echo/CMakeLists.txt:16]: Do the other examples have an install rule in their CMakeLists.txt? I'm used to having the option to run `make install` to put it in the Polysync install's bin directory so we might add that.

File: Echo/EchoNode.hpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+     * @brief Calls GetOpt::optionsParse. Standardizes cmd line handling.
+     * @param argv Argument vector.
+     * @param argc Argument count.
+     * @return
+     */
+    bool optionsParse( const int argc, char * argv[] );
+
+    /**
+     * @brief If help -h is one of arguments, don't start node up.
+     * @param argc
```
Comment (shnewto) [inline Echo/EchoNode.hpp]: Missing explanation of the params

File: Echo/PolySyncEchoHelp.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+{
+    std::vector< std::string > flagDescriptionsHelp;
+
+    flagDescriptionsHelp.emplace_back
+        ( " Show this help message [optional]." );
+
+    flagDescriptionsHelp.emplace_back
+        ( " Filter for a single message type [optional]. \n"
+        " Usage Example: $ polysync-echo -f ps_diagnostic_trace_msg \n\n"
+        " To see a list of message types: $ polysync-echo -f randomText \n\n"
```
Comment (shnewto) [inline Echo/PolySyncEchoHelp.cpp]: randomText? maybe this should read PolysyncMessageType?

File: Echo/PolySyncGetOpt.cpp [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        option_idx = getOptIdx( (const char) optret );
+
+        if ( option_idx == -1 )
+        {
+            cout << "\n\nUsage: invalid options. Usage guide follows." <<endl;
+
+            _getOptHelpFlag = true;
+        }
+
+         else if ( option_idx != -1 )
```
Comment (shnewto) [inline Echo/PolySyncGetOpt.cpp]: extra space before else?

File: Echo/EchoNode.hpp (hunk 2 above)
Comment (davepmiller) [inline Echo/EchoNode.hpp]: @return true if ?

File: Echo/CMakeLists.txt (hunk 1 above)
Comment (shnewto) [inline Echo/CMakeLists.txt:16]: It doesn't look like all the examples have install rules but I like the idea of having them. Adding the following to this file should do the trick but probably test it just to make sure.

install(
   TARGETS polysync-echo
   DESTINATION ${PSYNC_HOME}/bin
   )

File: Echo/EchoNode.cpp [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+#include <fstream>
+
+using namespace std;
+
+namespace polysync
+{
+
+void PolySyncEcho::initStateEvent()
+{
+    if ( getOpt.wasSingleMsgFiltered() )
```
Comment (davepmiller) [inline Echo/EchoNode.cpp]: Naming inconsistencies. We've got `msg` in some places and `message` in others. We usually prefer explicit naming.

Something like `getOpt.singleMessageFilter()`

File: Echo/EchoNode.cpp [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+
+void PolySyncEcho::registerSingleFilteredMessage()
+{
+    try
+    {
+        registerListener( getMessageTypeByName ( getOpt.getMsgName() ) );
+    }
+
```
Comment (davepmiller) [inline Echo/EchoNode.cpp]: Don't need the whitespace after the try block.

File: Echo/EchoNode.cpp [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+void PolySyncEcho::echoPolySyncMessagesToStdOut
+    ( std::shared_ptr < polysync:: Message > message )
+{
+    if ( getOpt.wereHeadersRequested() )
+    {
+        message->printHeader();
+    }
+
```
Comment (davepmiller) [inline Echo/EchoNode.cpp:82]: Unnecessary whitespace.

File: Echo/EchoNode.cpp [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    else if ( !getOpt.wereHeadersRequested() )
```
Comment (davepmiller) [inline Echo/EchoNode.cpp]: Just make this an `else`

File: Echo/PolySyncEchoHelp.hpp [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+ */
+namespace polysync
+{
+
+/**
+ * @brief The PolySyncEchoHelp class is a user interaction class to assist
+ *        end users with command line arguments/options for Echo utility.
+ *
+ */
+class PolySyncEchoHelp
```
Comment (davepmiller) [inline Echo/PolySyncEchoHelp.hpp]: If this object is already in the `polysync` namespace, no need to prefix it with `PolySync`.

Use-case == `polysync::EchoHelp` vs. `polysync::PolySyncEchoHelp`

File: Echo/PolySyncGetOpt.hpp [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+{
+
+/**
+ * @brief PolySyncGetOpt class
+ *
+ * The PolySyncGetOpt C++ class exists to build upon the C-based getopt class.
+ * Getopt parses command line input, and validates that input with options,
+ * arguments, and option-arguments.
+ */
+class PolySyncGetOpt
```
Comment (davepmiller) [inline Echo/PolySyncGetOpt.hpp]: Remove PolySync from object name.

Also, `GetOpt` isn't a very intuitive name. In a perfect world (such as ours) the name of this object would clearly reflect what it's responsible for. 

Possibilities:

1.) `ApplicationInputHandler`
2.) Ok.....#1 is all I could come up with off the bat.

File: Echo/PolySyncGetOpt.hpp [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+     *          or user entered invalid options/arguments on command line.
+     *
+     * If help flag true, PolySync nodes do not start up; help is displayed.
+     */
+    bool wasHelpRequested()
+    { return _getOptHelpFlag; }
+
+private:
+
+    char * _msgName;
```
Comment (davepmiller) [inline Echo/PolySyncGetOpt.hpp]: Use `std::string`

File: Echo/PolySyncGetOpt.cpp [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    else
+    {
+        return -1;
+    }
+}
+
+
+bool PolySyncGetOpt::optionsParse(const int argc, char *argv[])
+{
+    bool parsedOptSuccess = true;
```
Comment (davepmiller) [inline Echo/PolySyncGetOpt.cpp]: Could all of these variables live in a class?

File: Echo/PolySyncGetOpt.hpp [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+     * @param argv Argument vector.
+     * @param argc Argument count.
+     * @return
+     */
+    bool optionsParse( const int argc, char * argv[] );
+
+    /**
+     * @brief Member variable getter for message(s) name(s).
+     * @return Returns char * variable containing message name.
+     */
```
Comment (davepmiller) [inline Echo/PolySyncGetOpt.hpp]: 1.) Move these function definitions to the source file. 

2.) Always use `std::string` for strings.

3.) Getters should be `const`.  ---->   `std::string getMessageName() const;`

File: Echo/PolySyncGetOpt.cpp [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+         // success: container end not reached.
+    }
+    else
+    {
+        return -1;
+    }
+}
+
+
+bool PolySyncGetOpt::optionsParse(const int argc, char *argv[])
```
Comment (davepmiller) [inline Echo/PolySyncGetOpt.cpp]: This "arguments sandwich" needs "whitespace bread".

File: Echo/CMakeLists.txt (hunk 1 above)
Comment (abstractmachines) [inline Echo/CMakeLists.txt:16]: Improvements addressed and completed here: 
https://github.com/PolySync/PolySync-CPP-Examples/pull/7/commits/81a655e1d1a9ed1b22a1b6684daddeaabfef2a02

File: Echo/EchoNode.hpp (hunk 2 above)
Comment (abstractmachines) [inline Echo/EchoNode.hpp]: Improvements addressed + completed here: https://github.com/PolySync/PolySync-CPP-Examples/pull/7/commits/81a655e1d1a9ed1b22a1b6684daddeaabfef2a02

File: Echo/EchoNode.cpp (hunk 5 above)
Comment (abstractmachines) [inline Echo/EchoNode.cpp]: Improvements addressed + completed here:
https://github.com/PolySync/PolySync-CPP-Examples/pull/7/commits/81a655e1d1a9ed1b22a1b6684daddeaabfef2a02

File: Echo/EchoNode.cpp (hunk 8 above)
Comment (abstractmachines) [inline Echo/EchoNode.cpp]: Improvements addressed + completed here:
https://github.com/PolySync/PolySync-CPP-Examples/pull/7/commits/81a655e1d1a9ed1b22a1b6684daddeaabfef2a02

File: Echo/PolySyncEchoHelp.hpp (hunk 9 above)
Comment (abstractmachines) [inline Echo/PolySyncEchoHelp.hpp]: Improvements addressed + completed here:
https://github.com/PolySync/PolySync-CPP-Examples/pull/7/commits/81a655e1d1a9ed1b22a1b6684daddeaabfef2a02

File: Echo/PolySyncGetOpt.cpp (hunk 12 above)
Comment (abstractmachines) [inline Echo/PolySyncGetOpt.cpp]: Hi Dave, please clarify, we were unsure about this improvement suggestion

File: Echo/PolySyncGetOpt.hpp (hunk 13 above)
Comment (abstractmachines) [inline Echo/PolySyncGetOpt.hpp]: Improvements addressed + completed here:
https://github.com/PolySync/PolySync-CPP-Examples/pull/7/commits/81a655e1d1a9ed1b22a1b6684daddeaabfef2a02

File: Echo/PolySyncGetOpt.cpp (hunk 14 above)
Comment (abstractmachines) [inline Echo/PolySyncGetOpt.cpp]: Improvements addressed + completed here, let me know if this is better?

https://github.com/PolySync/PolySync-CPP-Examples/pull/7/commits/81a655e1d1a9ed1b22a1b6684daddeaabfef2a02

File: Echo/EchoNode.cpp (hunk 7 above)
Comment (abstractmachines) [inline Echo/EchoNode.cpp:82]: Improvements addressed + completed here:
https://github.com/PolySync/PolySync-CPP-Examples/pull/7/commits/81a655e1d1a9ed1b22a1b6684daddeaabfef2a02

File: Echo/EchoNode.cpp (hunk 8 above)
Comment (abstractmachines) [inline Echo/EchoNode.cpp]: Improvements addressed + completed here:
https://github.com/PolySync/PolySync-CPP-Examples/pull/7/commits/81a655e1d1a9ed1b22a1b6684daddeaabfef2a02

File: Echo/PolySyncEchoHelp.hpp (hunk 9 above)
Comment (abstractmachines) [inline Echo/PolySyncEchoHelp.hpp]: Improvements addressed + completed here:
https://github.com/PolySync/PolySync-CPP-Examples/pull/7/commits/81a655e1d1a9ed1b22a1b6684daddeaabfef2a02

File: Echo/PolySyncGetOpt.hpp (hunk 10 above)
Comment (abstractmachines) [inline Echo/PolySyncGetOpt.hpp]: Improvements addressed + completed here. I used ApplicationInputHandler.

https://github.com/PolySync/PolySync-CPP-Examples/pull/7/commits/81a655e1d1a9ed1b22a1b6684daddeaabfef2a02

File: Echo/PolySyncGetOpt.hpp (hunk 11 above)
Comment (abstractmachines) [inline Echo/PolySyncGetOpt.hpp]: Improvements addressed + completed here:
https://github.com/PolySync/PolySync-CPP-Examples/pull/7/commits/81a655e1d1a9ed1b22a1b6684daddeaabfef2a02

**Comment by shnewto** [comment]
Approved.
